### PR TITLE
Feature/tornado fetcher publisher

### PR DESCRIPTION
**Pull request by Dave-Haynes**
Status today:
* Unit tests run and pass locally and in CF.
* Service fails to load keys locally or in CF.
* Rabbit connection as yet untested in CF.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+language: python
+python:
+  - "3.5"
```
Comment (JamesGardiner) [inline .travis.yml]: Test against 3.6 as well - this is what we will use on CF

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+  - "3.4"
+sudo: required
+install: make depinstall
```
Comment (JamesGardiner) [inline .travis.yml]: This is failing in Travis

File: app/encrypter.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import base64
+import os
+
+import jwt
+from cryptography.hazmat.backends.openssl.backend import backend
+from cryptography.hazmat.primitives import serialization, hashes
+from cryptography.hazmat.primitives.asymmetric import padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+KID = 'SEFT'
```
Comment (JamesGardiner) [inline app/encrypter.py]: Will this be loaded from the environment eventually?

File: app/encrypter.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import base64
+import os
+
+import jwt
```
Comment (JamesGardiner) [inline app/encrypter.py]: Highly pedantic comment of the day - can we do imports alphabetically?

File: app/ftpclient.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+from ftplib import FTP
+from io import BytesIO
+import logging
+
+from publisher import Job
+
+
+class FTPWorker:
+
+    def __init__(self, user, password, host, port, timeout=30, **kwargs):
```
Comment (JamesGardiner) [inline app/ftpclient.py]: Is the kwarg  unpacking used at all? If not, can we just use explicit keyword params?

File: app/ftpclient.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        self.log = logging.getLogger("sdx.FTPWorker")
+        self.user, self.password = user, password
+        self.host, self.port = host, port
+        self.timeout = timeout
+        self.ftp = FTP()
+
+    def __enter__(self):
+        try:
+            self.ftp.connect(self.host, self.port, timeout=self.timeout)
+        except Exception as e:
```
Comment (JamesGardiner) [inline app/ftpclient.py:21]: Does this need to swallow all exceptions, or should we be more selective?

File: app/main.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        try:
+            with open(priv_fp, "r") as key_file:
+                priv_key = key_file.read()
+        except:
+            log.warning("Could not read key {0}".format(priv_fp))
+        try:
+            with open(pub_fp, "r") as key_file:
+                pub_key = key_file.read()
+        except:
+            log.warning("Could not read key {0}".format(pub_fp))
```
Comment (JamesGardiner) [inline app/main.py]: Might be useful to see the stacktrace of the original exception

File: app/main.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        help="Set a file path for log output")
+    return p
+
+
+def configure_log(args, name="sdx.seft"):
+    log = logging.getLogger(name)
+    log.setLevel(args.log_level)
+
+    formatter = logging.Formatter(
+        "%(asctime)s %(levelname)-7s %(name)s|%(message)s")
```
Comment (JamesGardiner) [inline app/main.py]: This format should be taken from the common library formatter

File: app/main.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    else:
+        fh = WatchedFileHandler(args.log_path)
+        fh.setLevel(args.log_level)
+        fh.setFormatter(formatter)
+        log.addHandler(fh)
+        ch.setLevel(logging.WARNING)
+
+    ch.setFormatter(formatter)
+    log.addHandler(ch)
+    log = logging.getLogger("pika")
```
Comment (JamesGardiner) [inline app/main.py]: again, this functionality is in the common library

File: app/main.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+    log.info(os.listdir("app/test"))
+    services = json.loads(os.getenv("VCAP_SERVICES", "{}"))
+
+    # Create the API service
+    app = make_app()
+    app.listen(args.port)
+
+    # Create the scheduled task
+    interval_ms = 30 * 60 * 1000
```
Comment (JamesGardiner) [inline app/main.py]: This should probably be configurable from the env

File: app/publisher.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+Job = namedtuple("Job", ["ts", "fn", "contents"])
+
+
+class DurableTopicPublisher:
+
+    EXCHANGE = "message"
+    PUBLISH_INTERVAL = 1
+    ROUTING_KEY = "example.text"
+
+    def __init__(self, amqp_url, queue_name, log=None, **kwargs):
```
Comment (JamesGardiner) [inline app/publisher.py]: Can we use explicit kwargs

File: app/test/decrypter.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
+import base64
+
+import jwt
```
Comment (JamesGardiner) [inline app/test/decrypter.py]: alphabetized

File: app/test/test_ftp.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+$ cf push seft-publisher-test
+$ cf logs seft-publisher-test --recent
+
+"""
+
+
+class NeedsTemporaryDirectory():
+
+    def setUp(self):
+        self.root = tempfile.mkdtemp()
```
Comment (JamesGardiner) [inline app/test/test_ftp.py:28]: In general, we should not be using the local filesystem in CF.

File: app/test/test_ftp.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+$ cf push seft-publisher-test
+$ cf logs seft-publisher-test --recent
+
+"""
+
+
+class NeedsTemporaryDirectory():
+
+    def setUp(self):
```
Comment (JamesGardiner) [inline app/test/test_ftp.py:27]: method names should comply with pep8

**Review by JamesGardiner — CHANGES_REQUESTED** [review]
Can you add a codecov.yml to this and set this up for codacy as well?

File: app/encrypter.py (hunk 3 above)
Comment (warrenbailey) [reply app/encrypter.py]: It should be configurable and once we move it to sdx-decrypt I think we do that, for now I'd be tempted to leave it as is.

File: app/main.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            "port": 2121,
+        }
+
+    def __init__(self, args, services):
+        self.args = args
+        self.services = services
+        self.publisher = DurableTopicPublisher(**self.amqp_params(services))
+
+    def transfer_files(self):
+        log = logging.getLogger("sdx.seft")
```
Comment (warrenbailey) [inline app/main.py]: Can the logger be defined at class level? Seems odd to redefine it every time we run transfer_files

File: app/main.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+def make_app():
+    return tornado.web.Application([
+        (r"/recent", StatusService, {"work": Task}),
+    ])
+
+
+def parser(description="SEFT Publisher service."):
+    here = os.path.dirname(__file__)
+    p = argparse.ArgumentParser(description)
+    p.add_argument(
```
Comment (warrenbailey) [inline app/main.py:137]: Why are we defining config via command line arguments? This goes against the other services that all use environment variables?

File: app/main.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+    log.setLevel(args.log_level)
+    log.addHandler(ch)
+    return name
+
+
+def main(args):
+    log = logging.getLogger(configure_log(args))
+    log.info("Launched in {0}.".format(os.getcwd()))
+
+    services = json.loads(os.getenv("VCAP_SERVICES", "{}"))
```
Comment (warrenbailey) [inline app/main.py]: This is specific to getting config from CF. I appreciate you're using CF for testing but production code will need  to be configured from environment variables as the target environment is NGD.

Therefore we should remove this line.

File: app/ftpclient.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+            self.log.warning(e)
+            return []
+
+    def get(self, filenames):
+        while filenames:
+            fp = filenames[0]
+            buf = BytesIO()
+            try:
+                self.ftp.retrbinary("RETR {0}".format(fp), callback=buf.write)
+            except Exception as e:
```
Comment (warrenbailey) [inline app/ftpclient.py:47]: Agree with James, can we avoid catching Exception and be a bit more fine grained?

File: app/main.py [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+
+
+class Task:
+    recent = OrderedDict([])
+
+    @staticmethod
+    def amqp_params(services):
+        log = logging.getLogger("sdx.seft.amqp")
+        try:
+            uri = services["rabbitmq"][0]["credentials"]["protocols"]["amqp"]["uri"]
```
Comment (warrenbailey) [inline app/main.py]: How will this work outside of CF?

**Review by warrenbailey — CHANGES_REQUESTED** [review]

File: app/test/test_ftp.py (hunk 14 above)
Comment (Dave-Haynes) [reply app/test/test_ftp.py:27]: Flake8 passes for me.

File: app/test/test_ftp.py (hunk 13 above)
Comment (Dave-Haynes) [reply app/test/test_ftp.py:28]: It's a test for an FTP server, so it needs a place to keep its files. I think that's legitimate in this case. If CF can't support that, its a deficiency in CF and someone should fix that.

File: app/main.py (hunk 16 above)
Comment (Dave-Haynes) [reply app/main.py:137]: Required for testing in CF. We don't need them in NGD.

File: app/ftpclient.py (hunk 18 above)
Comment (Dave-Haynes) [reply app/ftpclient.py:47]: There's no case I can think of where we could take a different action depending on the exception.

File: app/ftpclient.py (hunk 6 above)
Comment (Dave-Haynes) [reply app/ftpclient.py:21]: There may be different reasons for failure, but there isn't any choice of action we can take based on the type of exception. We either connect or we don't. If we don't, we can't proceed.

File: app/main.py [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+        worker = FTPWorker(**self.ftp_params(self.services))
+        encrypter = Encrypter(**self.encrypt_params(self.services, locn=self.args.keys))
+        with worker as active:
+            if not active:
+                return
+
+            for job in active.get(active.filenames):
+                if job.fn not in self.recent:
+                    data = job._asdict()
+                    data["contents"] = base64.standard_b64encode(job.contents).decode("ascii")
```
Comment (warrenbailey) [inline app/main.py]: Rather than "contents" it should be "file"
We also need a "filename" attribute with the name of the file.

**Review by warrenbailey — CHANGES_REQUESTED** [review]

File: Makefile [hunk 21]
```diff
@@ -0,0 +1,27 @@
+.PHONY: build dev start test
```
Comment (necrophonic) [inline Makefile]: Makefile doesn't work in clean environment - should be updated à la https://github.com/ONSdigital/sdx-receipt-ctp/blob/master/Makefile

File: README.md [hunk 22]
```diff
@@ -1,16 +1,29 @@
...
+## Getting started
+
+_TBD_
 
+## Configuration
 
-### Contributing
+| Environment variable | Default | Description
+| -------------------- | ------- | -----------
+| PORT                 | -       | The port to bind to
```
Comment (warrenbailey) [inline README.md]: Probably should add the environment variables to the README

File: app/ftpclient.py (hunk 18 above)
Comment (warrenbailey) [reply app/ftpclient.py:47]: OK

File: app/test/test_ftp.py [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+try:
+    from localserver import serve
+except ImportError:
+    # Travis CI
+    from app.test.localserver import serve
+
+"""
+To run this test in a Cloudfoundry environment:
+
+$ cf push seft-publisher-unittest
```
Comment (warrenbailey) [inline app/test/test_ftp.py]: This probably should be in the README

File: requirements.txt [hunk 24]
```diff
@@ -0,0 +1,5 @@
+cryptography==1.9
+tornado==4.5.1
+PyJWT==1.4.0
+pika==0.10.0
+sdx-common>=0.7.0
```
Comment (warrenbailey) [inline requirements.txt]: We specifically changed this in all other repos to pin the SDX common version

**Review by warrenbailey — CHANGES_REQUESTED** [review]

**Comment by warrenbailey** [comment]
FYI none of your commits are showing as verified, probably need to check your keys

**Review by warrenbailey — DISMISSED** [review]
Waiting for @JamesGardiner before merging

**Review by JamesGardiner — APPROVED** [review]

**Pull request merged**